xarray/coding/variables.py

```diff
@@ -232,20 +297,32 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype: np.typing.DTyp
     return data
 
 
-def _choose_float_dtype(dtype: np.dtype, has_offset: bool) -> type[np.floating[Any]]:
-    """Return a float dtype that can losslessly represent `dtype` values."""
-    # Keep float32 as-is.  Upcast half-precision to single-precision,
+def _choose_float_dtype(
```

Comment:
@mankoff do you have time to take a look here please?

Reply:
Hi @dcherian . I'm not sure what to look for. I will link to my open-but-stale PR that tried to start addressing this/similar issues: https://github.com/pydata/xarray/pull/6812

Perhaps also relevant are my last few comments on #2304 (see https://github.com/pydata/xarray/issues/2304#issuecomment-1201461626 ). The problem for me is that (1) the CF standards are ambiguously defined and (2) xarray needs to address the many use-cases where the CF standards are not followed (usually this means different data types).

Reply:
Thanks! Does this PR fix your original issue?

Reply:
Thanks @mankoff, I'll have a look at your extensive notes over there and try to come up with aomething.